Matlab Live Script, Assingment 2

sgjoshi25 opened the pull request:
This has a live script for the prelim statistical analysis on the experimental data provided.

sgjoshi25 committed:
Began working on experimental data

Began performing statistical analsis on the experimental data for the first cancer cell line.
Need to work on creating the heatmaps.

sgjoshi25 committed:
Merge branch 'master' of https://github.com/sriram-lab/metabolicmodeling into sgjoshi

sgjoshi25 committed:
Matlab Live Script, Assignment 2, statistical analysis on experimental data

THis is the statistical analysis on the experimental data for metabolism.

ScottCampit reviewed (changes requested):
Thanks, glad to see that you made use of the livescripts!

Note that .mlx files don't show the code inline when using the GitHub interface - sorry! If it's possible, please save the .mlx file as a .m, .html, and .pdf as well. I will share comments inline in the .m file, and can view figures as the html/pdf. Further, I moved your files into `/MATLAB/DFA/src/EDA`, as we're performing exploratory data analysis right now. For now, you can pull these changes `git pull origin sgjoshi` to view comments and edits I made in the livescript. 

Also, I'm evaluating your code as if we are preparing it for publication, and will suggest improvements in clarity, structure, and identify bugs if there are any. So sorry in advance if my comments are nit-picking, but it is with good intentions!  

**Major notes:** 
  * Please update the spreadsheet on Drive to reflect your sheets, or upload a .mat file of the data you're working on - I tried running your code but ran into errors when reading in the excel file. This will make it easier to evaluate the code. Alternatively, if you show the major data tables inline, that'll help.
  * For the `for loop` from line 10 to 17, the logic is really complicated. It would be hard for a novice coder to understand what's going on. I offer solutions to remedy this.
    * Additionally, make sure the averages you're taking are per condition. By iterating this for loops with a step of +1, you're actually taking averages across conditions as well, if the dataset contains each replicate as a column. To successfully get the averages you want, you want to add `n = n+2` at the end. That way, you don't need lines 13-16 either.
  * When I ran the code, I got a 3x3 correlation matrix with NaN (line 24). I offer a solution in the livescript, but it was because there are NaN values, which need to be turned to 0.
  * Your heatmaps, scatter plots, and coordinate plots look interesting, as we can see there are metabolites that are super high, and some that are super low. One thing we need to do is separate them by metabolite groups or metabolic pathways. 
     * Additionally, we probably need to scale the data, since the range of values is so large (0 to 1E8!). Try visualizing the data with the following transformations: 
        * min-max scaling (https://www.mathworks.com/help/matlab/ref/rescale.html)
        * mean 0 std 1 (https://www.mathworks.com/help/stats/zscore.html)
        * quantile normalization (https://www.mathworks.com/help/bioinfo/ref/quantilenorm.html)
        * log transformation (https://www.mathworks.com/help/matlab/ref/log.html) 

Very interesting results so far, and good job with your first iteration of exploratory data analysis! I'll have you implement these changes, and we'll review the code again later. 

I'll give you another assignment tomorrow, in addition to these changes.

Commit:
scampit reviewed livescript and moved scripts to MATLAB/DFA/src/EDA

ScottCampit commented on MATLAB/DFA/src/EDA/experimental_data.m:
```diff
@@ -0,0 +1,18 @@
+ exper_data = readtable('Barb Nelson Pos Table-20180410 - HL.xlsx', ...
```
Note that you can delete this file if you save `exper_data.mlx` as a .m file.

sgjoshi25 committed:
Added changes to the EDA figures

Made a more easy-to-read file.
Added data analysis to the second cancer cell line HCT116.
Fixed r matrix and NaN problems.

sgjoshi25 committed:
This is a .m file of the live script

sgjoshi25 commented:
Worked on fixing the mean and standard deviation tables and added tables for the HCT116 cancer cell line. Currently working on clustering the metabolomics into groups to better visualize trends.

ScottCampit commented on MATLAB/DFA/src/EDA/exper_data.m:
```diff
@@ -0,0 +1,131 @@
+%fill the tables with averages and standard deviations and then delete the
+%the last two trials for each condition
+for n = 2:13
+    Tabl_average_TU8902(:,(n)) = num2cell(nanmean(Tabl_average_TU8902{:,n:n+2},2));
```
unnecessary parentheses with `(n)`

ScottCampit commented on MATLAB/DFA/src/EDA/exper_data.m:
```diff
@@ -0,0 +1,131 @@
+    Tabl_average_HCT116(:,(n+2)) = [];
+    Tabl_std_HCT116(:,(n+2)) = [];
+    Tabl_average_HCT116(:,(n+1)) = [];
+    Tabl_std_HCT116(:,(n+1)) = [];
```
I don't think this for loop computes the values correctly still. 

If your columns are:
cond1 replicate 1 | cond1 replicate 2 | cond1 replicate 3 | cond2 replicate 1 ...

your for loop computes this:
i = 1: mean(cond1 replicate 1 | cond1 replicate 2 | cond1 replicate 3 | )
i = 2: mean(cond1 replicate 2 | cond1 replicate 3 | cond2 replicate 1 | )

and so on.

sgjoshi25 committed:
Merge branch 'master' of https://github.com/sriram-lab/metabolicmodeling into sgjoshi

sgjoshi25 committed:
Added changes to the foor loop

Fixed the for loop and checked it against a unit test. Deleted ducplicate code.

sgjoshi25 committed:
Created a function to calculate statistics.

This function will decrease the duplicate code in the original livescript. It contains the for loop and will create tables for mean and standard deviations of the given data.

sgjoshi25 committed:
Unit Test for the averaging functions for loop.

Tested the for loop and it seems to be working correctly with this unit test.

sgjoshi25 committed:
Parallel coordinate plots using the clutered metabolomics

Began adding the clustered metabolomics data to the script

sgjoshi25 committed:
The .m of the .mlx file updated

ScottCampit commented on MATLAB/DFA/src/EDA/exper_data.m:
```diff
@@ -0,0 +1,116 @@
+
+%% Create tables for averages and standard deviations. 
+%This here is going to take the averages and standard deviations from the original experimental data and fill two new tables with their values. Any NaN will be turned to 0 prior to finding the mean and standard deviations. 
+Tabl_average_TU8902 = table_calc(experimental_data_TU8902, 'mean');
```
Great, I saw your unittest and saw it passed. Also great that you created a function, especially if you repeat it several times.

ScottCampit commented on MATLAB/DFA/src/EDA/exper_data.m:
```diff
@@ -0,0 +1,116 @@
+%This here is going to take the averages and standard deviations from the original experimental data and fill two new tables with their values. Any NaN will be turned to 0 prior to finding the mean and standard deviations. 
+Tabl_average_TU8902 = table_calc(experimental_data_TU8902, 'mean');
+Tabl_average_TU8902(227,:) = [];
+Tabl_std_TU8902 = table_calc(experimental_data_TU8902, 'stddev');
```
I would suggest that you rename the function to something more descriptive, such as computeReplicateMetrics, just to make it clear it computes stuff like the mean, standard deviation, etc.

sgjoshi25 committed:
Added the clustering data

Created subplots of the mean and standard deviations to analyze the two cancer cell lines

sgjoshi25 committed:
Added changes to the .m file as well

sgjoshi25 committed:
Renamed the file to a more informative name

sgjoshi25 committed:
Added the NaN filter finder.

Tried to differentiate parallel coord plots for the different pathways but ran into bugs

sgjoshi25 committed:
Updated the .m file

sgjoshi25 committed:
Began adding the NaN Filter to the datasets

Need help deleting the metabolomics through the matlab code.

sgjoshi25 committed:
Updated the .m file

sgjoshi25 committed:
Worked on fixing the plots and the CV filter

Fixed the NaN filter. Need help exporting the tables as excel sheets from Matlab
Started the CV filter and created a function to decrease code duplication
Fixed the plots so that they output what is to be expected

sgjoshi25 committed:
Updated the .m file

sgjoshi25 committed:
Created a live function for the CV filter so that there is less code duplication

sgjoshi25 committed:
Created a function to be able to plot specific bio-groups of data

sgjoshi25 committed:
Added many changes to the EDA plots
Rotates the x-axis label and removed the underscores
separated the GOT1 and NT plots for the averages
scaled the datasets using normalize function between 0 and 1

sgjoshi25 committed:
Began updating the CV filter to better filter out the metabolomics

sgjoshi25 committed:
Fixed the CV filter to filter out the metabolomics.

Need help deleting the metabolites from the original tables.

sgjoshi25 committed:
Added NT vs. GOT1 plots to the coefficient of variation

sgjoshi25 committed:
updated the .m file

ScottCampit commented on MATLAB/DFA/src/EDA/exper_data.m:
```diff
@@ -0,0 +1,360 @@
+title 'Correlation Matrix of HCT116 mean values over 12 conditions';
+
+% create r matrix and heatmaps for the metabolomics
+corrmat = zeros(12);
```
I solved the issue with this bit of code: `corr` computes Pearson Correlation coefficients (by default) using the rows of your matrix. Instead of having your matrices be 12 x 226, have your matrices be 226 x 12. This will compute the metabolite-wise correlation coefficients.

An example snippet is shown below:
```corrmat = corr(mean_metabolomic_array_HCT116', mean_metabolomic_array_HCT116')```

ScottCampit commented on MATLAB/DFA/src/EDA/exper_data.m:
```diff
@@ -0,0 +1,360 @@
+end 
+% NOTE: Running this gives a matrix of NaNs.
+
+%% Create the correlation matrix, CV 
```
CV is Coefficient of Variation

ScottCampit commented on MATLAB/DFA/src/EDA/exper_data.m:
```diff
@@ -0,0 +1,360 @@
+CV_HCT116_GOT1 = std_HCT116_GOT1 ./ avg_HCT116_GOT1;
+CV_HCT116_NT = std_HCT116_NT ./ avg_HCT116_NT;
+
+%% Create heatmaps to visualize the averages
```
To remove stuff in front of the metabolite names:

```
metabolomics(:, 1) = strrep(metabolomics(:, 1), 'RP_Pos_', '');
metabolomics(:, 1) = strrep(metabolomics(:, 1), '_', '')
```

ScottCampit commented on MATLAB/DFA/src/EDA/exper_data.m:
```diff
@@ -0,0 +1,360 @@
+
+%% Create heatmaps to visualize the averages
+figure (3);
+heatmap(all_conditions, metabolomics(:,1),mean_metabolomic_array_TU8902, 'ColorScaling', 'log');
```
It may be easier to interpret after you do the NaN / CV filter. You don't have to visualize all datasets, just the one that contains ~60-70% of the dataset

ScottCampit commented on MATLAB/DFA/src/EDA/exper_data.m:
```diff
@@ -0,0 +1,360 @@
+
+aminoAcidIndex = findClusterType(metabolomics,'amino acid');
+subplot(3,1,1);
+parallelcoords(mean_metabolomic_array_TU8902(aminoAcidIndex,:), 'group', metabolomics(aminoAcidIndex,1), 'Labels', all_conditions);
```
change yaxis label to be either the "mean" or "standard deviation" instead of "Coordinate Value"

ScottCampit commented on MATLAB/DFA/src/EDA/exper_data.m:
```diff
@@ -0,0 +1,360 @@
+legend('location', 'bestoutside');
+hold off
+
+%% DFA Filter 1: CV thresholds
```
In practice, we'll want the dataset as well, not just the metabolite names, same with NaN filter

ScottCampit commented on MATLAB/DFA/src/EDA/exper_data.m:
```diff
@@ -0,0 +1,360 @@
+%% Use Filter 2 to create new data sheet
+%Now, we will use the Filter 2 to create an excel sheet for DFA. 
+%We will first locate all the locations that have only NaN's in tables and then delete the NaN rows. The HCT116 cell line should delete 7 rows while the TU8902 cell line should delete 6 rows. 
+for i = 1:219
```
So this doesn't work because across your 12 conditions, there are none that are all 0s. So the way I would implement this (for HCT116) is the following code:

```
% Construct an N x M boolean matrix that finds NaNs for each element
bool = isnan(table2array(Tabl_average_HCT116(:, 2:end)));

% If there are more than 6 empty values row-wise, then we'll remove the metabolite. Of course, if you find that there are still a lot NaNs, you can make this filter more stringent.
Tabl_average_HCT116(sum(double(bool), 2) >= 6, :) = [];
```

ScottCampit replied:
This also doesn't involve for loops, so the code is way more elegant - just two lines.

ScottCampit commented on MATLAB/DFA/src/EDA/exper_data.m:
```diff
@@ -0,0 +1,360 @@
+end 
+% Now, export the final table out as an excel sheet. 
+%{
+filename = 'Filter 2: HCT116 Cancer.xlsx';
```
So it was the filename that caused the issue: a best practice for filenames (especially when manipulating a lot of them), is to use underscores, and try not to use special symbols (ie ':') that have some programming meaning. Another way to name files is using something called camel case (ie fileName.xlsx)

ScottCampit replied:
'Sheet', 'SheetName'

sgjoshi25 committed:
Fixed the bugs in the CV filter so that it can work for any data inputed and any threshold specified

sgjoshi25 committed:
Changed the for loop to take in any size of metabolomics array such that EDA and plots can be created

sgjoshi25 committed:
Made this file easier to navigate
Only creates tables and filters the tables

sgjoshi25 committed:
updated the .m file

sgjoshi25 committed:
This file only creates the EDA heatmaps and correlation maps for the data presented

sgjoshi25 committed:
updated .m file

sgjoshi25 committed:
This script only created the parallel coordinate plots for EDA

sgjoshi25 committed:
updated .m file

ScottCampit commented on MATLAB/DFA/src/EDA/exper_data.m:
```diff
@@ -0,0 +1,228 @@
+CV_HCT116_GOT1 = std_HCT116_GOT1 ./ avg_HCT116_GOT1;
+CV_HCT116_NT = std_HCT116_NT ./ avg_HCT116_NT;
+
+%% DFA Filter 1: CV thresholds + NaN: 0.1
```
This bit here can be shortened using the following pseudocode:
cv = [0.1, 0.3, 0.6]
for i = 1:length(cv)
  Do all operations
  Store result in cell array
end

ScottCampit reviewed (approved):
For the most part, everything looks good! One minor thing improvement you can make is to shorten the CV filter section by saving the data in a cell array, then grabbing your data in the cell array for your plots. 

However, I'll approve this PR and leave it to you to make further improvements to this code and your future code as you continue.
